app/services/puppet_fact_parser.rb

```diff
@@ -201,37 +205,85 @@ def os_name
 
   def os_release
     case os_name
-    when /(suse|sles|gentoo)/i
-      facts[:operatingsystemrelease]
     when /(windows)/i
       facts[:kernelrelease]
     when /AIX/i
-      majoraix, tlaix, spaix, _yearaix = facts[:operatingsystemrelease].split("-")
+      majoraix, tlaix, spaix, _yearaix = os_release_full.split("-")
       majoraix + "." + tlaix + spaix
     when /JUNOS/i
-      majorjunos, minorjunos = facts[:operatingsystemrelease].split("R")
+      majorjunos, minorjunos = os_release_full.split("R")
       majorjunos + "." + minorjunos
     when /FreeBSD/i
-      facts[:operatingsystemrelease].gsub(/\-RELEASE\-p[0-9]+/, '')
+      os_release_full.gsub(/\-RELEASE\-p[0-9]+/, '')
     when /Solaris/i
-      facts[:operatingsystemrelease].gsub(/_u/, '.')
+      os_release_full.gsub(/_u/, '.')
     when /PSBM/i
-      majorpsbm, minorpsbm = facts[:operatingsystemrelease].split(".")
+      majorpsbm, minorpsbm = os_release_full.split(".")
       majorpsbm + "." + minorpsbm
     when /Archlinux/i
       # Archlinux is a rolling release, so it has no releases. 1.0 is always used
       '1.0'
     when /Debian/i
-      return "99" if facts[:lsbdistcodename] =~ /sid/
-      release = facts.dig(:os, :release, :full) || facts[:lsbdistrelease] || facts[:operatingsystemrelease]
+      return "99" if distro_codename =~ /sid/
+      release = os_release_full
       case release
       when 'bullseye/sid' # Debian Bullseye testing will be 11
         '11'
       else
         release
       end
     else
-      facts.dig(:os, :release, :full) || facts[:lsbdistrelease] || facts[:operatingsystemrelease]
+      os_release_full
     end
   end
+
+  def os_release_full
```

Comment:
should we have os_ and distro_ methods in abstract parser so all plugins parsers implement in a more standard way? I think you mentioned you wanted to do that, definitely not a must for this PR.

Reply:
That is indeed where I think we should go to. In https://github.com/theforeman/foreman/pull/8450 I did some of that work by using the minor releases from facter.